ethereum/contracts/Scale.sol

```diff
@@ -0,0 +1,86 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.6.2;
+
+contract Scale {
+
+    // Read a byte at a specific index and return it as type uint8
+    function readByteAtIndex(bytes memory data, uint8 index)
+        public
+        pure
+        returns (uint8)
+    {
+        return uint8(data[index]);
+    }
+
+    // Reverse uint8[] array in place
+    function reverse(uint8[] memory arr)
+        public
+        pure
+        returns(uint8[] memory)
+    {
+        for (uint i = 0; i < arr.length/2; i++) {
+            uint8 current = arr[i];
+            uint256 otherIndex = arr.length - i - 1;
+            arr[i] = arr[otherIndex];
+            arr[otherIndex] = current;
+        }
+        return arr;
+    }
+
+    // Convert an uint8[] array to uint256
+    function uint8ArrayToUint256(uint8[] memory arr)
+        public
+        pure
+        returns (uint256)
+    {
+        uint256 number;
+        for(uint i = 0; i < arr.length; i++){
+            number = number+uint(arr[i])*(2**(8*(arr.length-(i+1))));
+        }
+        return number;
+    }
+
+    // Decodes a SCALE encoded compact unsigned integer
+    function decodeUintCompact(bytes memory data)
+        public
+        pure
+        returns (uint256)
+    {
+        uint8 b = readByteAtIndex(data, 0);           // read the first byte
+        uint8 mode = b & 3;                           // bitwise operation
+
+        if(mode == 0) {
+            return b >> 2;                           // right shift to remove mode bits
+        } else if(mode == 1) {
+            uint8 bb = readByteAtIndex(data, 1);      // read the second byte
+            uint64 r = bb;                            // convert to uint64
+            r <<= 6;                                  // multiply by * 2^6
+            r += b >> 2;                              // right shift to remove mode bits
+            return r;
+        } else if(mode == 2) {
+            uint8 b2 = readByteAtIndex(data, 1);      // read the next 3 bytes
+            uint8 b3 = readByteAtIndex(data, 2);
+            uint8 b4 = readByteAtIndex(data, 3);
+
+            uint32 x1 = uint32(b) | uint32(b2) << 8;  // convert to little endian
+            uint32 x2 = x1 | uint32(b3) << 16;
+            uint32 x3 = x2 | uint32(b4) << 24;
+
+            x3 >>= 2;                                 // remove the last 2 mode bits
+            return uint256(x3);
+        } else if(mode == 3) {
+            uint8 l = b >> 2;                         // remove mode bits
+            require(l <= 63,                          // Max upper bound of 536 is (67 - 4)
+                    "Not supported: l>63 encountered when decoding a compact-encoded uint");
+            uint8[] memory buf = new uint8[](l+4);
+            for(uint8 i = 1; i < l+4+1; i++) {        // read bytes from data[1:l+4]
+                uint8 bNext = readByteAtIndex(data, i);
+                buf[i-1] = bNext;
```

Comment:
Instead of reversing below, we could also fill from back, right?

```Suggestion
                buf[l+4-i] = bNext;
```

Reply:
Good idea, this will definitely save some gas at runtime.

Reply:
If we remove the compact uint case 3 logic then both `uint8ArrayToUint256` and `reverse` can be removed. However, this suggestion still applies to the new method `reverseBytes`.

Before this PR is merged I'll either address it or make an issue for it.

Reply:
Created https://github.com/Snowfork/polkadot-ethereum/issues/56.